src/utils/Cypher.ts

```diff
@@ -26,19 +35,176 @@ export default class Cypher {
+          // Also log error?
+        }
+        cryptoSettingToSave = {
+          formerKey: keySettings.crypto.key,
+          key: configCryptoKey,
+          migrationDone: false
+        } as CryptoSetting;
+        isKeyChanged = true;
+      }
+    } else {
+      // Create New Config Crypto Key in Tenant Settings
+      cryptoSettingToSave = {
+        key: configCryptoKey
+      } as CryptoSetting;
+    }
+
+    // Key migration of senitive data
+    this.cryptoSetting = cryptoSettingToSave ? cryptoSettingToSave : keySettings.crypto;
+
+    if (cryptoSettingToSave) {
+      const keySettingToSave = {
+        id: keySettings?.id,
+        identifier: TenantComponents.CRYPTO,
+        type: CryptoSettingsType.CRYPTO,
+        crypto: cryptoSettingToSave
+      } as KeySettings;
+      await SettingStorage.saveCryptoSettings(tenantID, keySettingToSave);
+    }
+    return isKeyChanged;
+  }
+
+  public static async setMigrationDone(tenantID:string): Promise<void> {
+    this.cryptoSetting.migrationDone = true;
+    const keySettingToSave = {
+      id: this.keySetting.id,
+      identifier: TenantComponents.CRYPTO,
+      type: CryptoSettingsType.CRYPTO,
+      crypto: this.cryptoSetting
+    } as KeySettings;
+    await SettingStorage.saveCryptoSettings(tenantID, keySettingToSave);
+  }
+
+  public static getSensitiveDatainSetting(setting: SettingDB): string[] {
+
+    const allSensitiveData: string[] = [];
+    for (const property of setting.sensitiveData) {
+      // Check that the property does exist otherwise skip to the next property
+      if (_.has(setting, property)) {
+        const value:string = _.get(setting, property);
+
+        // If the value is undefined, null or empty then do nothing and skip to the next property
+        if (value && value.length > 0) {
+          allSensitiveData[`${property}`] = value;
+        }
+      }
+    }
+    return allSensitiveData;
+
+  }
+
+  public static async migrateSensitiveData(tenantID: string, setting: SettingDB): Promise<SensitiveData[]> {
+    try {
+      if (this.cryptoSetting) {
+        // TODO add check for clearValue
+
+        const valuebeforeDecrypt = this.getSensitiveDatainSetting(setting);
+        this.decryptSensitiveDataInJSON(setting, this.cryptoSetting.formerKey);
+        const valueAfterDecrypt = this.getSensitiveDatainSetting(setting);
+        this.encryptSensitiveDataInJSON(setting, this.cryptoSetting.key);
+        const valueAfterEncrypt = this.getSensitiveDatainSetting(setting);
+
+        const allSensitiveData: SensitiveData[] = [];
+        for (const property of setting.sensitiveData) {
+          const sensitiveData = {
+            identifier: setting.identifier,
+            path: property,
+            initialValue: {
+              encryptedValue: valuebeforeDecrypt[property],
+              key: this.cryptoSetting.formerKey
+            },
+            clearValue: valueAfterDecrypt[property],
+            migratedValue: {
```

Comment:
To test if a migration should be re-started, you can: 
+ add a reference in the settings to the array record used for current migration in sensitivedatamigration; 
+ test first for the migrationDone boolean;
+ check inside the sensitivedatamigration collection the matching array record if all array elements have initialValue and migratedValue fields filled with correct values.; 
+ run the migration again on missing array elements or not correctly migrated ones: test for migratedValue existence;
+ flag the migration has done.